File: 328-odd-even-linked-list/odd-even-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if(head == NULL || head->next == nullptr || head->next->next == NULL) return head;

        ListNode* odd = head;
        ListNode* even = head->next;
        ListNode* start = even;

        ListNode* temp = head->next->next;
        int cnt=3;

        while(temp){
            if(cnt & 1){
                // odd hai
                odd->next = temp;
                odd=temp;
            }
            else{
                // even hai
                even->next = temp;
                even=temp;
            }
            cnt++;
            temp = temp->next;
        }

        odd->next=start;
        even->next=NULL;
        return head;

    }
};

File: 328-odd-even-linked-list/odd-even-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode oddEvenList(ListNode head) {
        if(head == null || head.next == null || head.next.next == null) return head;

        ListNode odd = head;
        ListNode even = odd.next;
        ListNode currOdd = odd;
        ListNode currEven = even;

        while(currOdd != null && currEven != null){
            if(currEven != null){
                currOdd.next = currEven.next;
                currOdd = currOdd.next;
            }
            if(currOdd != null){
                currEven.next = currOdd.next;
                currEven = currEven.next;
            }
        }

            ListNode temp = odd;
            while(temp.next != null){
                temp = temp.next;
            }
            temp.next = even;

        return head;
    }
}